addons/product/models/product_template.py

```diff
@@ -513,67 +513,340 @@ def create_variant_ids(self):
         return True
 
     def has_dynamic_attributes(self):
-        return self.attribute_line_ids and \
-            any(create_variant == 'dynamic'
-            for create_variant in self.mapped('attribute_line_ids.value_ids.attribute_id.create_variant'))
+        """Return whether this `product.template` has at least one dynamic
+        attribute.
+
+        :return: True if at least one dynamic attribute, False otherwise
+        :rtype: bool
+        """
+        self.ensure_one()
+        return any(a.create_variant == 'dynamic' for a in self._get_valid_product_attributes())
 
     @api.multi
-    def get_filtered_variants(self, reference_product=None):
+    def _get_valid_product_template_attribute_lines(self):
+        """A product template attribute line is considered valid if it has at
+        least one possible value.
+
+        Those with only one value are considered valid, even though they should
+        not appear on the configurator itself (unless they have an is_custom
+        value to input), indeed single value attributes can be used to filter
+        products among others based on that attribute/value.
+
+        :return: all the valid product template attribute lines of this template
+        :rtype: recordset `product.template.attribute.line`
         """
-        Will filter availability (excluded) for the product
-        combinations (ex: color: white excludes size: large).
-
-        Will also filter availability (excluded) for the parent
-        product if specified (meaning that this product is a an optionnal or
-        accessory product of the reference_product).
-        Args:
-            reference_product (product.product): The reference product that has
-            the current product as an option or accessory product.
-        Returns:
-            The filtered list of product variants
+        self.ensure_one()
+        return self.attribute_line_ids.filtered(lambda ptal: ptal.product_template_value_ids)
```

Comment:
todo: probably better to replace .product_template_value_ids by .value_ids which is more efficient since the compute on .product_template_value_ids is not stored.

Reply:
And then the question (-> for master) becomes why is product_template_value_ids a computed many2many?

The ptal should be a many2one in ptav, and the reverse should be a standard non computed one2many.

Then product_tmpl_id can be a related in ptav, instead of being stored separately both in the ptal and ptav.
This removes the extra risk of them becoming unsync. And also remove the possibility for a ptav to exist without a matching ptal.

And value_ids in ptal should be a computed/related to the ptav.product_attribute_value_id. Then we create the inverse of the compute, instead of overriding create and write. For the sake of the backend interaface, maybe there are issue with this last suggestion.


This way we remove unnecessary duplicate fields and computes.